tools/RepositoryManagement/DevHome.RepositoryManagement/Services/RepositoryManagementDataAccessService.cs

```diff
@@ -0,0 +1,104 @@
+using DevHome.Database;
+using DevHome.Database.DatabaseModels.RepositoryManagement;
+using DevHome.RepositoryManagement.ViewModels;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Query;
+using Microsoft.Extensions.Hosting;
+using Serilog;
+
+namespace DevHome.RepositoryManagement.Services;
+
+public class RepositoryManagementDataAccessService
+{
+    private readonly ILogger _log = Log.ForContext("SourceContext", nameof(RepositoryManagementDataAccessService));
+
+    private readonly IHost _host;
+
+    public RepositoryManagementDataAccessService(IHost host)
+    {
+        // Store the host to make dbContext and RepositoryManagementItemViewModel
+        // dbContext can not be passed in because RepositoryManagementDataAccessService is singleton
+        // and dbContext needs to be scoped.
+        // RepositoryManagementItemViewModel can not be passed in because multiple are made.
+        // The best solution is to make factories for DevHomeContext and RepositoryManagementItemViewModel.
+        // Might be in a future change.
+        _host = host;
+    }
+
+    public void AddRepository(string repositoryName, string cloneLocation)
+    {
+        Repository newRepo = new();
+        newRepo.RepositoryName = repositoryName;
+        newRepo.RepositoryClonePath = cloneLocation;
+
+        RepositoryMetadata newMetadata = new();
+        newMetadata.Repository = newRepo;
+        newMetadata.RepositoryId = newRepo.RepositoryId;
+        newMetadata.IsHiddenFromPage = false;
+
+        newRepo.RepositoryMetadata = newMetadata;
+
+        var dbContext = _host.GetService<DevHomeDatabaseContext>();
+        dbContext.Add(newRepo);
+        dbContext.Add(newMetadata);
+        dbContext.SaveChanges();
+    }
+
+    public List<RepositoryManagementItemViewModel> GetRepositories(bool removeRepositoriesNotInTheirSavedLocation)
+    {
+        var repos = QueryDatabaseForRepositories(removeRepositoriesNotInTheirSavedLocation);
+        return ConvertToLineItems(repos);
+    }
+
+    private IIncludableQueryable<Repository, RepositoryMetadata> QueryDatabaseForRepositories(bool removeRepositoriesNotInTheirSavedLocation)
+    {
+        var dbContext = _host.GetService<DevHomeDatabaseContext>();
+        var repositoriesFromDatabase = dbContext.Repositories.Include(x => x.RepositoryMetadata);
+
+        if (!removeRepositoriesNotInTheirSavedLocation)
+        {
+            return repositoriesFromDatabase;
+        }
+
+        // The database can get out of sync.  Take this as an example.
+        // DevHome writes a repository.
+        // The user moves or deletes the repository.
+        // Now the database record has the correct name, but the incorrect location.
+        // Remove records where the directory does not exist.
+        repositoriesFromDatabase
+            .ToList() // EF does not understand Directory.Exists.  Fetch the data first.
+            .Where(x => !Directory.Exists(x.RepositoryClonePath))
```

Comment:
>!Directory.Exists(x.RepositoryClonePath) [](http://example.com/codeflow?start=24&length=40)

shouldn't this query for x.RepositoryClonePath/.git?

Reply:
The user has a folder that is designated as the location of a repository and, currently, it is.

Later in time.  The user removes the repository by removing everything under the path, but not the path.

Is that the scenario you are thinking of?